aequilibrae/project/data/matrices.py

```diff
@@ -0,0 +1,167 @@
+
+            if name in self.__items:
+                i = 0
+                while f'{name}_{i}' in self.__items:
+                    i += 1
+                name = f'{name}_{i}'
+            rec = self.create_record(name, fl)
+            rec.save()
+
+    def list(self) -> pd.DataFrame:
+        """
+
+        Returns:
+             df (:obj:`pd.DataFrame`:) Pandas DataFrame listing all matrices available in the model
+        """
+
+        def check_if_exists(file_name):
+            if os.path.isfile(os.path.join(self.fldr, file_name)):
+                return ''
+            else:
+                return 'file missing'
+
+        df = pd.read_sql_query('Select * from matrices;', self.conn)
+        df = df.assign(status='')
+        df.status = df.file_name.apply(check_if_exists)
+
+        return df
+
+    def get_matrix(self, matrix_name: str) -> AequilibraeMatrix:
+        """Returns an AequilibraE matrix available in the project
+
+        Raises an error if matrix does not exist
+
+        Args:
+            matrix_name (:obj:`str`:) Name of the matrix to be loaded
+
+        Returns:
+            matrix (:obj:`AequilibraeMatrix`:) Matrix object
+
+        """
+
+        return self.get_record(matrix_name).get_data()
+
+    def get_record(self, matrix_name: str) -> MatrixRecord:
+        """Returns a model Matrix Record for manipulation in memory"""
+
+        if matrix_name.lower() not in self.__items:
+            raise Exception('There is no matrix record with that name')
+
+        return self.__items[matrix_name.lower()]
+
+    def delete_record(self, matrix_name: str) -> None:
+        """Deletes a Matrix Record from the model and attempts to remove from disk"""
+        mr = self.get_record(matrix_name)
+        mr.delete()
+
+    def create_record(self, name: str, file_name: str) -> MatrixRecord:
+        """Creates a new record for a matrix in disk, but does not save it
+
+        If the matrix file is not already on disk, it will fail
+
+        Args:
+            *name* (:obj:`str`): Name of the matrix
+            *file_name* (:obj:`str`): Name of the file on disk
+
+        Return:
+            *matrix_record* (:obj:`MatrixRecord`): A matrix record that can be manipulated in memory before saving
+        """
+
+        if name in self.__items:
+            raise ValueError(f'There is already a matrix of name ({name}). It must be unique.')
+
+        for mat in self.__items.values():
+            if mat.file_name == file_name:
+                raise ValueError(f'There is already a matrix record for file name ({file_name}). It must be unique.')
+
+        tp = {key: None for key in self.__fields}
+        tp['name'] = name
+        tp['file_name'] = file_name
+        mat = AequilibraeMatrix()
+        mat.load(join(self.fldr, file_name))
+        tp['cores'] = mat.cores
+        mat.close()
+        del mat
+
+        mr = MatrixRecord(tp, self.__items, self.__project.logger)
```

Comment:
are you storing a name cache in the Matrices class and then passing down a reference to that to each instance of the Record class as well.

I feel that's not a good idea... you'd be better off enforcing that constraint at either the database level or in your Matrices class

Reply:
I had that feeling....

Reply:
Just addressed